scripts/features_extracting_training.py

```diff
@@ -0,0 +1,154 @@
+"""
+This file is used for:
+    1. run feature extractor
+    2. train the classifier
+on the exported spacer training data and confirms that
+it get similar (or higher) performance than production.
+"""
+
+import os
+import json
+import glob
+from io import BytesIO
+from boto.s3.key import Key
+
+from spacer import config
+from spacer.tasks import extract_features
+from spacer.data_classes import ImageLabels
+from spacer.train_classifier import trainer_factory
+from spacer.messages import DataLocation, ExtractFeaturesMsg
+
+
+class ExtractFeatures:
+    """
+    1. Initialize FeatureExtractor
+    2. Load images and anns.meta from s3
+    3. Extract features and save to local
+    """
+    def __init__(self, extractor_name):
+        self.extractor_name = extractor_name
+
+    def __call__(self, source_id, export_name='beta_export'):
+        conn = config.get_s3_conn()
+        bucket = conn.get_bucket('spacer-trainingdata', validate=True)
+
+        print("-> Fetching {} image and annotation meta files...".format(source_id))
+        all_keys = bucket.list(prefix='{}/s{}/images'.format(export_name, source_id))
+        img_keys = [key.name for key in all_keys if key.name.endswith('jpg')]
+        ann_keys = [Key(bucket=bucket, name=img.replace('jpg', 'anns.json')) for img in img_keys]
+
+        feats_local_path = os.path.join(config.LOCAL_FEATURE_DIR, 's{}'.format(source_id), 'images')
+        if not os.path.isdir(feats_local_path):
+            os.system('mkdir -p ' + feats_local_path)
+
+        print("-> Extracting features...")
+        for idx, im_key in enumerate(img_keys):
```

Comment:
Here, I suggest checking if the target `feature_loc` already exists on disk. If so, skip that iteration. This allows users to re-start the script if it hangs (mine did).

Reply:
updated